File: src/board.py
import random

import numpy as np


class Board:
    def __init__(self) -> None:
        self.new_game()

    def new_game(self) -> None:
        self.score = 0
        self.moves = 0
        self.board = np.array(
            [
                [0, 0, 0, 0],
                [0, 0, 0, 0],
                [0, 0, 0, 0],
                [0, 0, 0, 0],
            ]
        )

        for _ in range(2):
            pos = True
            while pos:
                i, j = random.sample(range(4), 2)
                pos = self.board[i][j]

            if random.randint(1, 10) == 1:
                self.board[i][j] = 4
            else:
                self.board[i][j] = 2

    def move(self, direction: str) -> bool:
        if direction == "down":
            self.board = np.flip(self.board)
            self.board = np.rot90(self.board)
            has_moved = self._moving()
            self.board = np.rot90(self.board, 3)
            self.board = np.flip(self.board)

        elif direction == "up":
            self.board = np.rot90(self.board)
            has_moved = self._moving()
            self.board = np.rot90(self.board, 3)

        elif direction == "right":
            self.board = np.fliplr(self.board)
            has_moved = self._moving()
            self.board = np.fliplr(self.board)

        elif direction == "left":
            has_moved = self._moving()

        if has_moved:
            self.moves += 1

        return has_moved

    def new_piece(self) -> bool:
        empty_positions = []
        for i in range(4):
            for j in range(4):
                if self.board[i][j] == 0:
                    empty_positions.append([i, j])

        if not empty_positions:
            return False

        i, j = random.choice(empty_positions)
        if random.randint(1, 10) == 10:
            self.board[i][j] = 4
        else:
            self.board[i][j] = 2

        return True

    def verify_end(self) -> bool:
        empty_positions = []
        for i in range(4):
            for j in range(4):
                if self.board[i][j] == 0:
                    empty_positions.append([i, j])

        if empty_positions:
            return False

        for i in range(4):
            for j in range(4):
                if j + 1 < 4:
                    if self.board[i][j] == self.board[i][j + 1]:
                        return False
                if j - 1 >= 0:
                    if self.board[i][j] == self.board[i][j - 1]:
                        return False
                if i + 1 < 4:
                    if self.board[i][j] == self.board[i + 1][j]:
                        return False
                if i - 1 >= 0:
                    if self.board[i][j] == self.board[i - 1][j]:
                        return False

        return True

    def _moving(self) -> bool:
        self.has_moved = False
        self._push()
        if self._sum():
            self._push()

        return self.has_moved

    def _push(self) -> None:
        for i in range(4):
            for j in range(1, 4):
                if self.board[i][j] != 0:
                    k = j
                    while k > 0 and self.board[i][k - 1] == 0:
                        self.board[i][k - 1], self.board[i][k] = (
                            self.board[i][k],
                            0,
                        )
                        self.has_moved = True
                        k -= 1

    def _sum(self) -> bool:
        result = False
        for i in range(4):
            for j in range(1, 4):
                if self.board[i][j] != 0:
                    if self.board[i][j - 1] == self.board[i][j]:
                        self.score += self.board[i][j] * 2
                        self.board[i][j - 1], self.board[i][j] = (
                            self.board[i][j] * 2,
                            0,
                        )
                        result = True
                        self.has_moved = True
        return result

File: src/rust-montecarlo/board.rs
use rand::{seq::SliceRandom, Rng};

pub struct Board {
    pub score: i32,
    pub moves: i32,
    pub board: [[i32; 4]; 4],
    has_moved: bool,
}

impl Board {
    pub fn new() -> Board {
        Board {
            score: 0,
            moves: 0,
            board: [[0; 4]; 4],
            has_moved: false,
        }
    }

    pub fn movement(&mut self, direction: &str) -> bool {
        self.has_moved = false;
        if direction == "down" {
            self.push_down();
            if self.sum_down() {
                self.push_down();
            }
        } else if direction == "up" {
            self.push_up();
            if self.sum_up() {
                self.push_up();
            }
        } else if direction == "right" {
            self.push_right();
            if self.sum_right() {
                self.push_right();
            }
        } else if direction == "left" {
            self.push_left();
            if self.sum_left() {
                self.push_left();
            }
        }

        if self.has_moved {
            self.moves += 1
        }
        self.has_moved
    }

    pub fn new_piece(&mut self) -> bool {
        let mut empty_positions: Vec<[i32; 2]> = Vec::new();

        for i in 0..4 {
            for j in 0..4 {
                if self.board[i][j] == 0 {
                    empty_positions.push([i as i32, j as i32]);
                }
            }
        }
        if empty_positions.len() > 0 {
            let pos = empty_positions.choose(&mut rand::thread_rng()).unwrap();

            let i = pos[0];
            let j = pos[1];

            if rand::thread_rng().gen_range(1..=10) == 10 {
                self.board[i as usize][j as usize] = 4;
            } else {
                self.board[i as usize][j as usize] = 2;
            }
            true
        } else {
            false
        }
    }

    pub fn verify_end(&self) -> bool {
        for i in 0..4 {
            for j in 0..4 {
                if self.board[i][j] == 0 {
                    return false;
                } else if j < 3 && self.board[i][j] == self.board[i][j + 1] {
                    return false;
                } else if j > 0 && self.board[i][j] == self.board[i][j - 1] {
                    return false;
                } else if i < 3 && self.board[i][j] == self.board[i + 1][j] {
                    return false;
                } else if i > 0 && self.board[i][j] == self.board[i - 1][j] {
                    return false;
                }
            }
        }
        true
    }

    fn push_left(&mut self) {
        for i in 0..4 {
            for j in 1..4 {
                if self.board[i][j] != 0 {
                    let mut k = j;
                    while k > 0 && self.board[i][k - 1] == 0 {
                        (self.board[i][k - 1], self.board[i][k]) = (self.board[i][k], 0);
                        self.has_moved = true;
                        k -= 1;
                    }
                }
            }
        }
    }

    fn push_up(&mut self) {
        for i in 1..4 {
            for j in 0..4 {
                if self.board[i][j] != 0 {
                    let mut k = i;
                    while k > 0 && self.board[k - 1][j] == 0 {
                        (self.board[k - 1][j], self.board[k][j]) = (self.board[k][j], 0);
                        self.has_moved = true;
                        k -= 1;
                    }
                }
            }
        }
    }

    fn push_right(&mut self) {
        for i in 0..4 {
            for j in (0..3).rev() {
                if self.board[i][j] != 0 {
                    let mut k = j;
                    while k < 3 && self.board[i][k + 1] == 0 {
                        (self.board[i][k + 1], self.board[i][k]) = (self.board[i][k], 0);
                        self.has_moved = true;
                        k += 1;
                    }
                }
            }
        }
    }

    fn push_down(&mut self) {
        for i in (0..3).rev() {
            for j in 0..4 {
                if self.board[i][j] != 0 {
                    let mut k = i;
                    while k < 3 && self.board[k + 1][j] == 0 {
                        (self.board[k + 1][j], self.board[k][j]) = (self.board[k][j], 0);
                        self.has_moved = true;
                        k += 1;
                    }
                }
            }
        }
    }

    fn sum_left(&mut self) -> bool {
        let mut result = false;
        for i in 0..4 {
            for j in 1..4 {
                if self.board[i][j] != 0 {
                    if self.board[i][j - 1] == self.board[i][j] {
                        self.score += self.board[i][j] * 2;
                        (self.board[i][j - 1], self.board[i][j]) = (self.board[i][j] * 2, 0);
                        result = true;
                        self.has_moved = true;
                    }
                }
            }
        }
        result
    }

    fn sum_up(&mut self) -> bool {
        let mut result = false;
        for i in 1..4 {
            for j in 0..4 {
                if self.board[i][j] != 0 {
                    if self.board[i - 1][j] == self.board[i][j] {
                        self.score += self.board[i][j] * 2;
                        (self.board[i - 1][j], self.board[i][j]) = (self.board[i][j] * 2, 0);
                        result = true;
                        self.has_moved = true;
                    }
                }
            }
        }
        result
    }

    fn sum_right(&mut self) -> bool {
        let mut result = false;
        for i in 0..4 {
            for j in (0..3).rev() {
                if self.board[i][j] != 0 {
                    if self.board[i][j + 1] == self.board[i][j] {
                        self.score += self.board[i][j] * 2;
                        (self.board[i][j + 1], self.board[i][j]) = (self.board[i][j] * 2, 0);
                        result = true;
                        self.has_moved = true;
                    }
                }
            }
        }
        result
    }

    fn sum_down(&mut self) -> bool {
        let mut result = false;
        for i in (0..3).rev() {
            for j in 0..4 {
                if self.board[i][j] != 0 {
                    if self.board[i + 1][j] == self.board[i][j] {
                        self.score += self.board[i][j] * 2;
                        (self.board[i + 1][j], self.board[i][j]) = (self.board[i][j] * 2, 0);
                        result = true;
                        self.has_moved = true;
                    }
                }
            }
        }
        result
    }
}
